node/network/protocol/src/lib.rs

```diff
@@ -612,67 +613,46 @@ pub mod vstaging {
-		pub candidate_hash: CandidateHash,
-		/// Validator that attested the validity.
-		pub signed_by: ValidatorIndex,
-		/// Signature of seconding validator.
-		pub signature: ValidatorSignature,
-	}
+		Statement(Hash, UncheckedSignedStatement),
 
-	impl StatementDistributionMessage {
-		/// Get fingerprint describing the contained statement uniquely.
-		pub fn get_fingerprint(&self) -> (CompactStatement, ValidatorIndex) {
-			match self {
-				Self::Statement(_, statement) => (
-					statement.unchecked_payload().to_compact(),
-					statement.unchecked_validator_index(),
-				),
-				Self::LargeStatement(meta) =>
-					(CompactStatement::Seconded(meta.candidate_hash), meta.signed_by),
-			}
-		}
-
-		/// Get the signature from the statement.
-		pub fn get_signature(&self) -> ValidatorSignature {
-			match self {
-				Self::Statement(_, statement) => statement.unchecked_signature().clone(),
-				Self::LargeStatement(metadata) => metadata.signature.clone(),
-			}
-		}
-
-		/// Get contained relay parent.
-		pub fn get_relay_parent(&self) -> Hash {
-			match self {
-				Self::Statement(r, _) => *r,
-				Self::LargeStatement(meta) => meta.relay_parent,
-			}
-		}
-
-		/// Whether this message contains a large statement.
-		pub fn is_large_statement(&self) -> bool {
-			if let Self::LargeStatement(_) = self {
-				true
-			} else {
-				false
-			}
-		}
+		/// A notification of a backed candidate being known by the
+		/// sending node, for the purpose of being requested by the receiving node
+		/// if needed.
+		#[codec(index = 1)]
+		BackedCandidateInv {
+			/// The relay-parent of the candidate.
+			relay_parent: Hash,
+			/// The hash of the candidate.
+			candidate_hash: CandidateHash,
+			/// The para that the candidate is assigned to.
+			para_id: ParaId,
+			/// The head-data corresponding to the candidate.
+			parent_head_data_hash: Hash,
+			/// A bitfield which indicates which validators in the para's
+			/// group at the relay-parent have seconded this candidate.
+			///
+			/// This MUST have the minimum amount of bytes
+			/// necessary to represent the number of validators in the
+			/// assigned backing group as-of the relay-parent.
+			seconded_in_group: BitVec<u8, bitvec::order::Lsb0>,
+		},
+
+		/// A notification of a backed candidate being known by the sending known,
+		/// for the purpose of informing a receiving node which already has the candidate.
+		#[codec(index = 2)]
+		BackedCandidateKnown(Hash, CandidateHash),
+
+		/// All messages for V1 for compatibility with the statement distribution
+		/// protocol, for relay-parents that don't support asynchronous backing.
+		///
+		/// These are illegal to send to V1 peers, and illegal to send concerning relay-parents
+		/// which support asynchronous backing. This backwards compatibility should be
+		/// considered immediately deprecated and can be removed once the node software
+		/// is not required to support asynchronous backing anymore.
```

Comment:
Ok, we do this because once we negotiated v2 with a peer, we would still like to handle all those still existing old leaves, otherwise we could not do that once we switched protocols. For v1 only peers, we keep using v1 anyway.

Reply:
Exactly

Reply:
The v2 protocol in principle also would work for old-style leaves without AB. But then we'd have an issue of fragmentation in the network between v1 and v2 peers, leading to liveness failures unless we introduced some code on v2 to bridge the data from the v1 networking logic to the v2 networking logic, which is extra complexity that is unnecessary.